### ISSUE TITLE
Shared accounts for stat159

### ISSUE THREAD

**Issue by ryanlovett** [berkeley-dsep-infra/hubs#4267]
### Bug description

The stat159 deployment has RTC enabled and it seems to be working well. @fperez would like to take this a step further and isolate storage. Right now one student can access another student's data for the duration of the shared session.

We imagine shared accounts defined in the hub config, and not provided by the authenticator (because special purpose accounts are too difficult to create at scale in instruction). The config would also scope which users have authority to act as those shared accounts.

```
- shared-1:
  - user1
  - user2
- shared-2:
  - user2
  - user3
  - user4
 ```

It would make sense to define the above relationships using JupyterHub groups.

Given the above `user3` would authenticate as themselves and start a server as `shared-2`, and the hub would mount `shared-2`'s home directory. There is no UI for this -- the user would have to be educated about specifying the URL.

I think the hard part is getting the [scopes](https://jupyterhub.readthedocs.io/en/stable/rbac/scopes.html#) right.

In our setting, we would want to make sure that the shared user account names are not present in our authenticator. I doubt there are CalNet shortnames of the form `shared-N`, but it'd be good to determine a secure convention.

### Environment & setup

- Hub: stat159

### How to reproduce

 .

**Comment by balajialg** [comment]
@ryanlovett If I understand this right, the action to be performed is to make config changes related to scopes in datahub config files to support this enhancement. Is this something an urgent need that you (or) anyone assigned will enable during this sprint or should this get scoped as a backlog?

**Comment by ryanlovett** [comment]
@fperez was hoping to have something ready in a couple of weeks. I'll give it a shot! (I assigned myself.)

**Comment by fperez** [comment]
Huge thanks!! I wanted to tag @minrk @rcthomas @shreddd on this one: it gives us a chance to experiment with some of our discussions on the NERSC side. I'll be using this in practice immediately with my students, and we're going for a very similar model of shared accounts, so I wanted to both make sure you were aware of this, and also perhaps to get some input on any useful tips you may have on the config tests you showed recently, that we at the Berkeley team could learn from?

@minrk  - could I ask you to point me again to the config example you showed?

**Comment by fperez** [comment]
Tagging also @facusapienza21 and @isaacdsloan so you're in the loop as we test this model for shared projects.

**Comment by fperez** [comment]
Thinking out loud about naming conventions for these shared accounts (since I presume we need to know that they don't collide with calnet ones?), perhaps something like `<HUBNAME>-shared.<ANYTHING>`? Where the last part allows instructors to define their own patterns (things like `proj1-group1`, or `hw3-g3`, etc.).

**Comment by minrk** [comment]
@fperez it's [here](https://github.com/minrk/scirahub-config/blob/main/collaborative/jupyterhub_config.py). The key points to keep in mind:

1. roles and role assignments are static, and require Hub restart to reload config if they should change
2. groups and group assignments are not, and can be updated via APIs or authenticator-managed groups (e.g. from GitHub team membership)

So for each shared account, I create:

1. a user (does not need to be able to login, but Spawner does need to be able to launch it)
2. a role granting access to that user's server
3. a group with that role

Then you can grant and revoke user permissions to the shared server at any time via group membership changes, but you'll need to update config and relaunch the server to add or remove a collaboration account.

I wrote the sample config with a [`projects` dict](https://github.com/minrk/scirahub-config/blob/0d3e95817bc09ea3703039dc71ddda9d28078e7b/collaborative/jupyterhub_config.py#L49-L57) representing the collaborations, but I imagine a production deployment would get this info from either a separately managed file or external API.

**Comment by fperez** [comment]
Fantastic, extremely useful @minrk, huge thanks!! @ryanlovett pls LMK if you'd like to chat about this, I'm very keen on helping this move forward.

Tagging also @yuvipanda @consideRatio @tsnow03 so we can keep an eye on this and start considering how to replicate it as a starting point in the JMTE/CryoCloud hubs. We'll likely want these things elsewhere, but those seem like good places to start dog-fooding with controlled communities.

**Comment by ryanlovett** [comment]
Thanks @minrk ! This looks straightforward.

@fperez One complicating factor on our side is that [our authenticator](https://github.com/berkeley-dsep-infra/canvasoauthenticator) is set up to [manage groups](https://github.com/berkeley-dsep-infra/datahub/blob/staging/hub/values.yaml#L202), so we can't define them in configuration in this way for most of our hubs. Right now our authenticator prepares groups based on course affiliations. This is useful in that we can do things like allocate resources based on enrollments. For a course-specific hub like stat159, this isn't necessary, and for this hub we can configure the authenticator to not manage groups.

For our other hubs, when we want to roll this out more broadly, we'll need another way. It appears that bCourses (Canvas) lets instructors manage groups within their course (append `/groups` to a course's bcourses URL). I think, in parallel to this particular issue, that it'd be good if the Canvas authenticator could also consume groups from the instructor-managed group feature of Canvas.

**Comment by ryanlovett** [comment]
@minrk I configured a hub with something very similar to yours (see below at bottom), however the spawner immediately fails to launch servers for the collab users (regular users are fine).  It raises:
```
[E 2023-02-21 21:41:46.278 JupyterHub user:832] Unhandled error starting collab-test1's server: not enough values to unpack (expected 3, got 1)
    Traceback (most recent call last):
      File "/usr/local/lib/python3.11/site-packages/jupyterhub/user.py", line 747, in spawn
        url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<string>", line 113, in start
    ValueError: not enough values to unpack (expected 3, got 1)
```

I can create other new users in /hub/admin and start their servers without trouble. I can also create new groups. The error is only present when the users and groups are associated via the roles.

We are using `jupyterhub/k8s-hub:2.0.1-0.dev.git.5959.h65c52001`. I can file a new issue with jupyterhub or KubeSpawner if you think this warrants it.

(currently in this configuration, all users have the rtc cli param added, though we may switch to using something like your `collaborative` group when this goes to production)
z2jh config snippet:
```
# manage_groups is false

jupyterhub:
    hub:
        extraConfig:
            # Use 0x-<title> in `values.yaml` extraConfig values.
            # We want these to come *before* the extraConfig values
            # in `values.yaml`. Since these are ordered alphabetically,
            # 1x-<title> used in `common.yaml` will come after
            # 0x-<title> used in `values.yaml` - so config set there
            # will override config set here.
            20-roles-and-shared-accounts: |
                collab_groups = {
                    "test1": [
                        "rylo",
                        "scf-stat",
                        "fernando.perez",
                        "fsapienza",
                    ],
                    "test2": [
                        "rylo",
                        "balajialwar",
                        "fernando.perez",
                        "fsapienza",
                    ],
                }

                for collab_group, members in collab_groups.items():
                    print(f"Add group '{collab_group}' with members: {members}")
                    c.JupyterHub.load_groups[collab_group] = members

                    # Add user associated with collab group
                    collab_user = f"collab-{collab_group}"

                    # Anyone who in a collab_group can admin and access the
                    # user server, and can list users of that group.
                    c.JupyterHub.load_roles.append(
                        {
                            "name": f"collab-access-{collab_group}",
                            "scopes": [
                                "admin-ui",
                                f"list:users!user={collab_user}",
                                f"admin:servers!user={collab_user}",
                                f"access:servers!user={collab_user}",
                            ],
                            "groups": [collab_group],
                        }
                    )
```

If I delete the `load_roles.append()`, redeploy, then delete and recreate the test groups and associated `collab-` users, I can start the user servers. I think this essentially clears out the roles that were loaded previously.

**Comment by minrk** [comment]
Thanks for testing, that's not a very good traceback. What Spawner do you have, and is there perhaps an overridden `.start()`? That traceback is pretty weird and doesn't show where the error really comes from, but it sounds like a `.start()` method that might be defined in a config file. Could that be the case?

**Comment by ryanlovett** [comment]
Thanks @minrk, you were right. The exception was in custom groups handling code in our overridden `.start()` method. Sorry to bother!

**Comment by ryanlovett** [comment]
This is working, although there is an outstanding issue where the hub raises an exception if we try to load groups from the configuration. (despite disabling group management in the Authenticator)

**Comment by fperez** [comment]
Thx @ryanlovett! Is that remaining issue being tracked elsewhere already or known/reported as a core JupyterHub issue?

**Comment by ryanlovett** [comment]
@fperez No, I haven't filed it upstream. I have a feeling it is due to some local config but I'm not sure where.

My guess is in how we set [c.Authenticator.manage_groups](https://jupyterhub.readthedocs.io/en/stable/reference/authenticators.html#authenticator-managed-group-membership) because that is the only think I know of that would cause this behavior. Our other hubs get group info from the Canvas authenticator, but we have to turn it off in stat159 to make our own groups. When we define groups in config, the hub tells us we're not allowed to create groups, although we can do it manually in the admin page. I might start a discussion on discourse.

**Comment by ryanlovett** [comment]
@fperez This was an issue in the local config, and I'll patch it shortly.
